Iceberg query failure because of predicate pushdown with iceberg column id

**Context**

> we have some Iceberg tables where we manually generate the parquet data files and Iceberg metadata ourselves. 


When `id` for the type is missing from the parquet file, this can cause a query failure on the call:

https://github.com/trinodb/trino/blob/7190f200962dc77d0ccb11a8e34beb71865e43b0/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java#L937


**Relevant stack trace**
```
Query 20240609_161836_02951_zpmtj failed: Error opening Iceberg split /path/data/file.parquet (offset=0, length=1660): Cannot invoke "org.apache.parquet.schema.Type$ID.intValue()" because the return value of "org.apache.parquet.schema.PrimitiveType.getId()" is null
io.trino.spi.TrinoException: Error opening Iceberg split /mnt/shavast01_datalake/iceberg/temp/_managedtmp/6p7pl0cwmpdw/iceberg_caches/iceberg-temp-ghjkloyvmfircvttdsyufowgjqtmgzwh/data/file.parquet (offset=0, length=1660): Cannot invoke "org.apache.parquet.schema.Type$ID.intValue()" because the return value of "org.apache.parquet.schema.PrimitiveType.getId()" is null
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.createParquetPageSource(IcebergPageSourceProvider.java:1132)
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.createDataPageSource(IcebergPageSourceProvider.java:633)
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.createPageSource(IcebergPageSourceProvider.java:373)
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.createPageSource(IcebergPageSourceProvider.java:265)
        at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorPageSourceProvider.createPageSource(ClassLoaderSafeConnectorPageSourceProvider.java:48)
        at io.trino.split.PageSourceManager.createPageSource(PageSourceManager.java:61)
        at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:264)
        at io.trino.operator.Driver.processInternal(Driver.java:403)
        at io.trino.operator.Driver.lambda$process$8(Driver.java:306)
        at io.trino.operator.Driver.tryWithLock(Driver.java:709)
        at io.trino.operator.Driver.process(Driver.java:298)
        at io.trino.operator.Driver.processForDuration(Driver.java:269)
        at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:890)
        at io.trino.execution.executor.dedicated.SplitProcessor.run(SplitProcessor.java:77)
        at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.lambda$run$0(TaskEntry.java:191)
        at io.trino.$gen.Trino_448____20240607_201000_2.run(Unknown Source)
        at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.run(TaskEntry.java:192)
        at io.trino.execution.executor.scheduler.FairScheduler.runTask(FairScheduler.java:168)
        at io.trino.execution.executor.scheduler.FairScheduler.lambda$submit$0(FairScheduler.java:155)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.parquet.schema.Type$ID.intValue()" because the return value of "org.apache.parquet.schema.PrimitiveType.getId()" is null
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.lambda$getParquetTupleDomain$35(IcebergPageSourceProvider.java:1504)
        at com.google.common.collect.CollectCollectors.lambda$toImmutableMap$7(CollectCollectors.java:195)
        at java.base/java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1787)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:546)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:702)
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.getParquetTupleDomain(IcebergPageSourceProvider.java:1504)
        at io.trino.plugin.iceberg.IcebergPageSourceProvider.createParquetPageSource(IcebergPageSourceProvider.java:1017)
        ... 25 more
```

**Slack discussion**

https://trinodb.slack.com/archives/CJ6UC075E/p1717777005901119

**Technical notes**

The issue reported here is likely related to https://github.com/trinodb/trino/pull/19066

Avoid doing the predicate pushdown when any of the descriptors from the parquet schema is missing the **id** value.

For the corresponding PR - add an integration test to ensure we'll not be dealing with further regressions (either through hive migrate procedure or potentially https://iceberg.apache.org/docs/latest/spark-procedures/#add_files)

--- a/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java
+++ b/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java
@@ -1421,6 +1421,7 @@ public class IcebergPageSourceProvider
         }
 
         Map<Integer, ColumnDescriptor> descriptorsById = descriptorsByPath.values().stream()
+                .filter(descriptor -> descriptor.getPrimitiveType().getId() != null)
                 .collect(toImmutableMap(descriptor -> descriptor.getPrimitiveType().getId().intValue(), identity()));
         ImmutableMap.Builder<ColumnDescriptor, Domain> predicate = ImmutableMap.builder();
         effectivePredicate.getDomains().orElseThrow().forEach((columnHandle, domain) -> {

--- a/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java
+++ b/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergV2.java
@@ -16,9 +16,13 @@ package io.trino.plugin.iceberg;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import io.trino.Session;
+import io.trino.filesystem.FileEntry;
+import io.trino.filesystem.FileIterator;
+import io.trino.filesystem.Location;
 import io.trino.filesystem.TrinoFileSystemFactory;
 import io.trino.plugin.base.util.Closables;
 import io.trino.plugin.blackhole.BlackHolePlugin;
+import io.trino.plugin.hive.TestingHivePlugin;
 import io.trino.plugin.hive.metastore.HiveMetastore;
 import io.trino.plugin.hive.metastore.HiveMetastoreFactory;
 import io.trino.plugin.iceberg.fileio.ForwardingFileIo;
@@ -87,6 +91,7 @@ import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.concurrent.Executors.newFixedThreadPool;
 import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.apache.iceberg.FileFormat.ORC;
+import static org.apache.iceberg.FileFormat.PARQUET;
 import static org.apache.iceberg.TableProperties.SPLIT_SIZE;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
@@ -109,6 +114,11 @@ public class TestIcebergV2
         metastore = ((IcebergConnector) queryRunner.getCoordinator().getConnector(ICEBERG_CATALOG)).getInjector()
                 .getInstance(HiveMetastoreFactory.class)
                 .createMetastore(Optional.empty());
+
+        queryRunner.installPlugin(new TestingHivePlugin(queryRunner.getCoordinator().getBaseDataDir().resolve("iceberg_data")));
+        queryRunner.createCatalog("hive", "hive", ImmutableMap.<String, String>builder()
+                .put("hive.security", "allow-all")
+                .buildOrThrow());
 
         try {
             queryRunner.installPlugin(new BlackHolePlugin());
@@ -227,6 +237,45 @@ public class TestIcebergV2
         Table icebergTable = loadTable(tableName);
         writeEqualityDeleteToNationTable(icebergTable, Optional.of(icebergTable.spec()), Optional.of(new PartitionData(new Long[] {1L})));
         assertQuery("SELECT array_column[1], map_column[1], row_column.x FROM " + tableName, "SELECT 1, 2, 1 FROM nation WHERE regionkey != 1");
+    }
+
+    @Test
+    public void testParquetMissingFieldId()
+            throws Exception
+    {
+        String hiveTableName = "test_hive_parquet" + randomNameSuffix();
+        assertUpdate("CREATE TABLE hive.tpch." + hiveTableName + "(names ARRAY(varchar)) WITH (format = 'PARQUET')");
+        assertUpdate("INSERT INTO hive.tpch." + hiveTableName + " VALUES ARRAY['Alice', 'Bob'], ARRAY['Carol', 'Dave'], ARRAY['Eve', 'Frank']", 3);
+
+        String location = metastore.getTable("tpch", hiveTableName).orElseThrow().getStorage().getLocation();
+        FileIterator files = fileSystemFactory.create(SESSION).listFiles(Location.of(location));
+        ImmutableList.Builder<FileEntry> fileEntries = ImmutableList.builder();
+        while (files.hasNext()) {
+            FileEntry file = files.next();
+            if (!file.location().path().contains(".trinoSchema") && !file.location().path().contains(".trinoPermissions")) {
+                fileEntries.add(file);
+            }
+        }
+        FileEntry parquetFile = getOnlyElement(fileEntries.build());
+
+        String icebergTableName = "test_iceberg_parquet" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + icebergTableName + "(names ARRAY(varchar))");
+        Table icebergTable = loadTable(icebergTableName);
+        icebergTable.newAppend()
+                .appendFile(DataFiles.builder(icebergTable.spec())
+                        .withPath(parquetFile.location().toString())
+                        .withFileSizeInBytes(parquetFile.length())
+                        .withRecordCount(3)
+                        .withFormat(PARQUET).build())
+                .commit();
+        icebergTable.updateProperties()
+                .set("schema.name-mapping.default", "[{\"field-id\":1,\"names\":[\"names\"]}]")
+                .commit();
+
+        assertQuery("SELECT * FROM " + icebergTableName, "VALUES ARRAY['Alice', 'Bob'], ARRAY['Carol', 'Dave'], ARRAY['Eve', 'Frank']");
+
+        assertUpdate("DROP TABLE hive.tpch." + hiveTableName);
+        assertUpdate("DROP TABLE " + icebergTableName);
     }
 
     @Test

